<a href="https://colab.research.google.com/github/nssn96/ML_iris/blob/main/ML_iris.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# ML Assignmet 1 - Iris dataset

In [190]:
#import statements
from sklearn.datasets import load_iris
import numpy as np
import pandas as pd
import matplotlib.pyplot as plt
import itertools

from sklearn.model_selection import train_test_split
from sklearn.linear_model import LinearRegression
from sklearn.metrics import mean_squared_error
%matplotlib inline

In [191]:
#loading the iris dataset
iris_dataset=load_iris()
print(type(iris_dataset))

<class 'sklearn.utils.Bunch'>


In [192]:
#to explore the iris_dataset dictionary and understand the different available keys and thier values
print(iris_dataset.keys())
#print(iris_dataset.data)
print(iris_dataset.target)
print(iris_dataset.frame)
print(iris_dataset.target_names)
print(iris_dataset.feature_names)

dict_keys(['data', 'target', 'frame', 'target_names', 'DESCR', 'feature_names', 'filename', 'data_module'])
[0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0
 0 0 0 0 0 0 0 0 0 0 0 0 0 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1
 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 2 2 2 2 2 2 2 2 2 2 2
 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2
 2 2]
None
['setosa' 'versicolor' 'virginica']
['sepal length (cm)', 'sepal width (cm)', 'petal length (cm)', 'petal width (cm)']


From the above code snippet we can see that we have 

*   Four input features, which are sepal length, sepal width, petal length, petal width.
*   Three target names depending on the input features. The target names are : setosa=0, versicolor=1,virginica=2



In [193]:
#storing features and target of iris dataset separately
x = pd.DataFrame(iris_dataset.data)  # x is in matrix format
y=pd.DataFrame(iris_dataset.target)   # y is in a vector format
x.columns=['sepal_length','sepal_width','petal_length','petal_width']
y.columns=['target']

In [194]:
#to check the number/size of the data matrix
print(x.shape)

(150, 4)


So, from above code we can see that we have a total of 150 samples of input features

In [195]:
#converting to dataframe
data = pd.DataFrame(data= np.c_[iris_dataset['data'], iris_dataset['target']],columns= iris_dataset['feature_names'] + ['target'])
data['species'] = pd.Categorical.from_codes(iris_dataset.target,iris_dataset.target_names)
data.columns=['sepal_length','sepal_width','petal_length','petal_width','target','species']

In [196]:

# plt.scatter(data.sepal_length,data.sepal_width)
# plt.xlabel("Sepal length")
# plt.ylabel("Sepal Width")
# plt.show()


In [199]:
#splitting the data into train and test set
x_train,x_test,y_train,y_test = train_test_split(x,y,test_size=0.1, random_state=1,stratify = y)

In [ ]:
# x1=x_train.petal_length
# y1 = x_train.petal_width
# num_obs = len(x1)
# noise = np.random.normal(loc=0.0, scale=5.0, size=num_obs)
# y= 10+(2*x1)+noise

# plt.figure(figsize=(20,7))
# plt.plot(x1, y, 'o')


In [200]:
#class definition for Linear regression
class LinearRegression:

  def fit(self,data,target,steps=100):
    alpha=0.01 #learning rate
    self.m = data.shape #num of training examples,features
    self.data=data
    self.target=target
    self.bias=0
    self.w=np.zeros(self.m)
    loss_arr=[]

    for i in range(steps):
      target_pred = self.predict(self.data)


      # calculate gradient descent
      dbias = -2 * np.sum(self.target-target_pred) / self.m
      dw = -(2 * (self.data.T).dot(self.target-target_pred)) / self.m

      #least square function
      #mse_loss =  (np.sum((self.target-target_pred) ** 2)) / self.m
      #mse_loss = np.square(np.subtract(self.target,target_pred)).mean()
      mse_loss = mean_squared_error(self.target,target_pred)
      loss_arr.append(mse_loss)

      #update weights
      self.bias-= (alpha * dbias)
      self.w -= (alpha * dw)

    return loss_arr

  def predict(self,data):
    return data.dot(self.w) + self.bias


In [ ]:
# # 12 diff combinations of models
# model = LinearRegression()
# print(len(x1))
# l  = model.fit(x1,y1)
# print(l)
# # step_num = range(1,101)
# # plt.plot(step_num, l, label='loss vs step')

In [201]:
# 12 diff combinations of models
model = LinearRegression()
temp = ['sepal_length','sepal_width','petal_length','petal_width']
li = list(itertools.permutations(temp,2))

for i in li:
  feature1=i[0]
  feature2=i[1]
  x1 = x_train[feature1]
  y1 = x_train[feature2]
  l  = model.fit(x1,y1)
  print(l)


TypeError: ignored

In [ ]:


# # fitting the model with data set

# # reg_sgd = SGDRegressor(loss='squared_error', max_iter=100)
# # reg_sgd.fit(X,y)
# # y_hat_pred = reg_sgd.predict(X_test)
# backup = data



# MSE_score_train = []
# MSE_score_test = []

# mod_no = range(1,13)
# for i in mod_no:
#   model = SGDRegressor(loss="squared_error", max_iter=100)
#   model.fit(x_train, y_train)
#   y_train_hat  = model.predict(x_train)
#   y_test_hat  = model.predict(x_test)
#   MSE_score_train.append(-1 * mean_squared_error(y_train,y_train_hat))
#   MSE_score_test.append(-1 * mean_squared_error(y_test,y_test_hat))
    
# plt.plot(mod_no,  MSE_score_train, label='train')
# plt.plot(mod_no,  MSE_score_test, label='test')
# plt.xlabel("Model Number")
# plt.ylabel("MSE")
# plt.legend()

# # plt.annotate('Sweet Spot', xy = (9, -.005), xytext=(9, -.015),
# #             arrowprops = dict(facecolor='black', shrink = 0.05));